src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/ExecutionManager/Helpers/NibbleMapHelpers.cs

```diff
@@ -0,0 +1,147 @@
+    // the effective address
+    internal const ulong BytesPerBucket = 8 * MapUnit.SizeInBytes;
+
+    // We load the map contents as 32-bit integers, which contains 8 4-bit nibbles.
+    // The algorithm will focus on each nibble in a map unit before moving on to the previous map unit
+    internal readonly struct MapUnit
+    {
+        public const int SizeInBytes = sizeof(uint);
+        public const ulong SizeInNibbles = 2 * SizeInBytes;
+        public readonly uint Value;
+
+        public MapUnit(uint value) => Value = value;
+
+        public override string ToString() => $"0x{Value:x}";
+
+        // Shift the next nibble into the least significant position.
+        public MapUnit ShiftNextNibble => new MapUnit(Value >>> 4);
+
+        public const uint NibbleMask = 0x0Fu;
+        internal Nibble Nibble => new((int)(Value & NibbleMask));
+
+        public bool IsZero => Value == 0;
+
+        // Assuming mapIdx is the index of a nibble within the current map unit,
+        // shift the unit so that nibble is in the least significant position and return the result.
+        public MapUnit FocusOnIndexedNibble(MapKey mapIdx)
+        {
+            int shift = mapIdx.GetNibbleShift();
+            return new MapUnit (Value >>> shift);
+        }
+    }
+
+    // Each nibble is a 4-bit integer that gives an offset within a bucket.
+    // We reserse 0 to mean that there is no method starting at any offset within a bucket
+    internal readonly struct Nibble
+    {
+        public readonly int Value;
+
+        public Nibble(int value)
+        {
+            Debug.Assert (value >= 0 && value <= 0xF);
+            Value = value;
+        }
+
+        public static Nibble Zero => new Nibble(0);
+        public bool IsEmpty => Value == 0;
+
+        public ulong TargetByteOffset
+        {
+            get
+            {
+                Debug.Assert(Value != 0);
+                return (uint)(Value - 1) * MapUnit.SizeInBytes;
+            }
+        }
+    }
+
+    // The key to the map is the index of an individual nibble
+    internal readonly struct MapKey
+    {
+        private readonly ulong MapIdx;
+        public MapKey(ulong mapIdx) => MapIdx = mapIdx;
+        public override string ToString() => $"0x{MapIdx:x}";
+
+        // The offset of the address in the target space that this map index represents
+        public ulong TargetByteOffset => MapIdx * BytesPerBucket;
+
+        // The index of the map unit that contains this map index
+        public ulong ContainingMapUnitIndex => MapIdx / MapUnit.SizeInNibbles;
+
+        // The offset of the map unit that contains this map index
+        public ulong ContainingMapUnitByteOffset => ContainingMapUnitIndex * MapUnit.SizeInBytes;
+
+        // The map index is the index of a nibble within the map, this gives the index of that nibble within a map unit.
+        public int NibbleIndexInMapUnit => (int)(MapIdx & (MapUnit.SizeInNibbles - 1));
```

Comment:
Since this is going to be a computed index, we should either convert this to `uint` or at least assert in the getter the result is >= 0.

Reply:
Changed to a `uint`, this value should always be `[0,7]`